File: agentmem/sdk/python/examples/recorder_control_quickstart.py
"""Synthetic end-to-end Universal Recorder + Gate + remediation example.

This example prints only identifiers, readiness, and dispositions. It never
prints credentials or captured payloads.
"""

from __future__ import annotations

import hashlib
import json
import os
from datetime import UTC, datetime
from uuid import uuid4

from lians import (
    LiansClient,
    a2a_event,
    lians_event,
    mcp_jsonrpc_event,
    otlp_genai_span,
)


def main() -> None:
    base_url = os.getenv("LIANS_URL", "http://localhost:8000")
    api_key = os.environ.get("LIANS_API_KEY", "")
    access_token = os.environ.get("LIANS_ACCESS_TOKEN", "")
    mediator_api_key = os.environ.get("LIANS_MEDIATOR_API_KEY", "")
    if not api_key and not access_token:
        raise SystemExit("Set LIANS_API_KEY or LIANS_ACCESS_TOKEN (the value is never printed).")
    if mediator_api_key and api_key and mediator_api_key == api_key:
        raise SystemExit("LIANS_MEDIATOR_API_KEY must be a separate credential.")

    now = datetime.now(UTC)
    suffix = uuid4().hex[:8]
    run_id = f"synthetic-order-review-{suffix}"
    trace_id = "0123456789abcdef0123456789abcdef"
    span_id = "0123456789abcdef"
    events = [
        lians_event(
            "decision.started",
            {
                "name": "synthetic-order-review",
                "phase": "started",
                "model_id": "synthetic-model",
                "policy_version": "quickstart-1",
                "input": {"order_id": "SYNTHETIC-001", "amount": 125},
                "evidence": ["synthetic-source:catalog-v1"],
            },
            run_id=run_id,
            trace_id=trace_id,
            span_id=span_id,
            agent_id="synthetic-review-agent",
            principal_id="synthetic-workload",
            occurred_at=now,
            idempotency_key=f"{run_id}:decision-started",
        ),
        otlp_genai_span(
            trace_id=trace_id,
            span_id="1123456789abcdef",
            parent_span_id=span_id,
            operation="chat",
            model="synthetic-model",
            input=[{"role": "user", "content": "Review synthetic order SYNTHETIC-001"}],
            output=[{"role": "assistant", "content": "Synthetic review completed"}],
            agent_id="synthetic-review-agent",
            run_id=run_id,
            occurred_at=now,
            ended_at=now,
        ),
        mcp_jsonrpc_event(
            {
                "jsonrpc": "2.0",
                "id": "synthetic-tool-call-1",
                "method": "tools/call",
                "params": {
                    "name": "synthetic_catalog_lookup",
                    "arguments": {"sku": "SYNTHETIC-SKU"},
                },
            },
            run_id=run_id,
            tool_name="synthetic_catalog_lookup",
            agent_id="synthetic-review-agent",
            occurred_at=now,
        ),
        a2a_event(
            {
                "kind": "task",
                "id": "synthetic-a2a-task-1",
                "contextId": run_id,
                "status": {"state": "completed", "timestamp": now.isoformat()},
                "artifacts": [{"name": "synthetic-review", "parts": [{"text": "approved"}]}],
            },
            run_id=run_id,
            agent_id="synthetic-review-agent",
        ),
    ]

    with LiansClient(
        base_url=base_url,
        api_key=api_key,
        access_token=access_token,
    ) as client:
        principal = client.whoami()
        if mediator_api_key:
            with LiansClient(base_url=base_url, api_key=mediator_api_key) as mediator:
                mediator_principal_id = mediator.whoami()["principal_id"]
        else:
            # This must be the canonical whoami principal of a separate
            # broker/sidecar credential. The zero UUID is intentionally
            # unredeemable when no mediator credential was configured.
            mediator_principal_id = os.getenv(
                "LIANS_MEDIATOR_PRINCIPAL_REF",
                "lians:principal:v1:api-key:00000000-0000-0000-0000-000000000000",
            )
        if mediator_principal_id == principal["principal_id"]:
            raise SystemExit("The evaluator and mediator must be separate identities.")
        capabilities = client.platform_capabilities()
        platform = client.platform_readiness()
        print(
            "platform",
            {
                "status": platform["status"],
                "recorder_version": capabilities["components"]["recorder"]["version"],
            },
        )
        batch = client.ingest_recorder_batch(events)
        recorder_run_id = batch["results"][0]["readiness"]["run_id"]
        readiness = client.recorder_run_readiness(recorder_run_id)
        print(
            "recorder",
            {
                "accepted": batch["accepted"],
                "score": readiness["score"],
                "ready": readiness["receipt_ready"],
            },
        )

        policy = client.create_gate_policy(
            {
                "name": f"synthetic-release-gate-{suffix}",
                "version": "quickstart-1",
                "default_disposition": "deny",
                "protected_actions": ["synthetic.order.release"],
                "target_ref_prefixes": ["urn:lians:synthetic-order:"],
                "enforcement_principal_ids": [mediator_principal_id],
                "maximum_permit_ttl_seconds": 30,
                "rules": [
                    {
                        "name": "require-recorded-policy",
                        "applies_to_risk_levels": ["high", "critical"],
                        "require_policy_attached": True,
                        "action_on_failure": "deny",
                    }
                ],
                "metadata": {"synthetic": True},
            }
        )
        policy = client.activate_gate_policy(policy["id"])
        decision = client.record_decision(
            agent_id="synthetic-review-agent",
            decision_type="order_release",
            outcome="approved",
            decided_at=now,
            policy_version="quickstart-1",
            metadata={"risk_level": "high", "synthetic": True},
        )
        provider_request = {
            "action": "synthetic.order.release",
            "decision_id": decision["id"],
            "target_ref": "urn:lians:synthetic-order:SYNTHETIC-001",
            "arguments": {"synthetic": True, "order_id": "SYNTHETIC-001"},
        }
        execution_request_hash = hashlib.sha256(
            json.dumps(
                provider_request,
                sort_keys=True,
                separators=(",", ":"),
                ensure_ascii=False,
            ).encode("utf-8")
        ).hexdigest()
        verdict = client.evaluate_gate(
            {
                "action": "synthetic.order.release",
                "target_ref": "urn:lians:synthetic-order:SYNTHETIC-001",
                "decision_id": decision["id"],
                "enforcement_principal_id": mediator_principal_id,
                "permit_ttl_seconds": 30,
                "execution_request_hash": execution_request_hash,
                "risk_level": "high",
                "policy_set_id": policy["id"],
                "context": {"synthetic": True, "recorder_run_id": recorder_run_id},
            }
        )
        permit = verdict.get("execution_permit")
        permit_consumed = False
        if permit is not None and mediator_api_key:
            # A real mediator derives these claims from the normalized request
            # it is about to dispatch and recomputes the digest independently.
            mediator_request_hash = hashlib.sha256(
                json.dumps(
                    provider_request,
                    sort_keys=True,
                    separators=(",", ":"),
                    ensure_ascii=False,
                ).encode("utf-8")
            ).hexdigest()
            with LiansClient(base_url=base_url, api_key=mediator_api_key) as mediator:
                mediator.consume_gate_execution_permit(
                    {
                        "permit_id": permit["permit_id"],
                        "token": permit["token"],
                        "action": provider_request["action"],
                        "target_ref": provider_request["target_ref"],
                        "decision_id": provider_request["decision_id"],
                        "execution_request_hash": mediator_request_hash,
                    }
                )
            permit_consumed = True
        print(
            "gate",
            {
                "principal": principal["principal_id"],
                "disposition": verdict["disposition"],
                "permit_id": permit["permit_id"] if permit is not None else None,
                "permit_consumed": permit_consumed,
            },
        )

        case = client.create_investigation_case(
            {
                "title": "Synthetic control-plane exercise",
                "description": "Synthetic data only; validates owned remediation flow.",
                "severity": "low",
                "gate_decision_id": verdict["id"],
                "metadata": {"synthetic": True},
            }
        )
        task = client.create_remediation_task(
            case["id"],
            {
                "expected_case_updated_at": case["updated_at"],
                "title": "Attest synthetic evidence review",
                "description": "No customer or production data is used.",
                "owner_principal": principal["principal_id"],
                "metadata": {"synthetic": True},
            },
        )
        task_in_progress = client.update_remediation_task(
            task["id"],
            {
                "expected_updated_at": task["updated_at"],
                "status": "in_progress",
            },
        )
        task_closed = client.close_remediation_task(
            task["id"],
            {
                "expected_updated_at": task_in_progress["updated_at"],
                "statement": "Synthetic evidence review completed.",
                "evidence_refs": [f"gate:{verdict['id']}", f"recorder-run:{recorder_run_id}"],
            },
        )
        case_before_close = client.investigation_case(case["id"])
        case_closed = client.close_investigation_case(
            case["id"],
            {
                "expected_updated_at": case_before_close["updated_at"],
                "statement": "Synthetic exercise closed after the owned task was attested.",
                "evidence_refs": [f"attestation:{task_closed['attestation']['id']}"],
                "resolution_summary": "Synthetic control-plane path completed.",
            },
        )
        print(
            "investigation",
            {"case_id": case["id"], "status": case_closed["status"]},
        )


if __name__ == "__main__":
    main()

File: agentmem/sdk/typescript/examples/recorder-control-quickstart.ts
/** Synthetic Universal Recorder + Gate + attested-remediation workflow. */

import { createHash } from "node:crypto";

import {
  LiansClient,
  a2aEvent,
  liansEvent,
  mcpJsonRpcEvent,
  otlpGenAiSpan,
} from "../src/index.js";

const baseUrl = process.env.LIANS_URL ?? "http://localhost:8000";
const apiKey = process.env.LIANS_API_KEY;
const accessToken = process.env.LIANS_ACCESS_TOKEN;
const mediatorApiKey = process.env.LIANS_MEDIATOR_API_KEY;
// Without a separate mediator credential the zero-UUID fallback is
// intentionally unredeemable; the evaluator never consumes its own permit.
const configuredMediatorPrincipalId =
  process.env.LIANS_MEDIATOR_PRINCIPAL_REF ??
  "lians:principal:v1:api-key:00000000-0000-0000-0000-000000000000";
if (!apiKey && !accessToken) {
  throw new Error("Set LIANS_API_KEY or LIANS_ACCESS_TOKEN; credentials are never printed.");
}
if (mediatorApiKey && apiKey && mediatorApiKey === apiKey) {
  throw new Error("LIANS_MEDIATOR_API_KEY must be a separate credential.");
}

function canonicalJson(value: unknown): string {
  if (value === null || typeof value !== "object") return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(canonicalJson).join(",")}]`;
  const object = value as Record<string, unknown>;
  return `{${Object.keys(object)
    .sort()
    .map((key) => `${JSON.stringify(key)}:${canonicalJson(object[key])}`)
    .join(",")}}`;
}

async function main(): Promise<void> {
const suffix = Date.now().toString(36);
const runId = `synthetic-ts-review-${suffix}`;
const traceId = "fedcba9876543210fedcba9876543210";
const occurredAt = new Date();
const client = new LiansClient({ baseUrl, apiKey, accessToken });

const events = [
  liansEvent(
    "decision.started",
    {
      name: "synthetic-ts-review",
      phase: "started",
      model_id: "synthetic-model",
      policy_version: "quickstart-ts-1",
      input: { order_id: "SYNTHETIC-TS-001" },
      evidence: ["synthetic-source:catalog-v1"],
    },
    {
      runId,
      traceId,
      spanId: "fedcba9876543210",
      agentId: "synthetic-ts-agent",
      occurredAt,
      idempotencyKey: `${runId}:decision-started`,
    },
  ),
  otlpGenAiSpan({
    runId,
    traceId,
    spanId: "edcba98765432101",
    operation: "chat",
    model: "synthetic-model",
    input: [{ role: "user", content: "Review a synthetic order" }],
    output: [{ role: "assistant", content: "Synthetic review complete" }],
    agentId: "synthetic-ts-agent",
    occurredAt,
    endedAt: occurredAt,
  }),
  mcpJsonRpcEvent(
    {
      jsonrpc: "2.0",
      id: "synthetic-ts-tool-1",
      method: "tools/call",
      params: { name: "synthetic_lookup", arguments: { sku: "SYNTHETIC-SKU" } },
    },
    { runId, toolName: "synthetic_lookup", agentId: "synthetic-ts-agent", occurredAt },
  ),
  a2aEvent(
    {
      kind: "task",
      id: "synthetic-ts-task-1",
      contextId: runId,
      status: { state: "completed", timestamp: occurredAt.toISOString() },
      artifacts: [{ name: "synthetic-review", parts: [{ text: "approved" }] }],
    },
    { runId, agentId: "synthetic-ts-agent" },
  ),
];

const principal = await client.whoami();
let mediatorPrincipalId = configuredMediatorPrincipalId;
if (mediatorApiKey) {
  const mediatorIdentity = await new LiansClient({
    baseUrl,
    apiKey: mediatorApiKey,
  }).whoami();
  mediatorPrincipalId = mediatorIdentity.principal_id ?? "";
}
if (
  !principal.principal_id ||
  !mediatorPrincipalId ||
  mediatorPrincipalId === principal.principal_id
) {
  throw new Error("The evaluator and mediator must be separate identities.");
}
const capabilities = await client.platformCapabilities();
const platform = await client.platformReadiness();
console.log("platform", {
  status: platform.status,
  recorder_version: capabilities.components.recorder?.version,
});
const batch = await client.ingestRecorderBatch(events);
const recorderRunId = batch.results[0]?.readiness.run_id;
if (!recorderRunId) throw new Error("Recorder accepted no events");
const readiness = await client.recorderRunReadiness(recorderRunId);
console.log("recorder", { accepted: batch.accepted, score: readiness.score, ready: readiness.receipt_ready });

let policy = await client.createGatePolicy({
  name: `synthetic-ts-release-gate-${suffix}`,
  version: "quickstart-ts-1",
  default_disposition: "deny",
  protected_actions: ["synthetic.order.release"],
  target_ref_prefixes: ["urn:lians:synthetic-order:"],
  enforcement_principal_ids: [mediatorPrincipalId],
  maximum_permit_ttl_seconds: 30,
  rules: [{
    name: "require-recorded-policy",
    applies_to_risk_levels: ["high", "critical"],
    require_policy_attached: true,
    action_on_failure: "deny",
  }],
  metadata: { synthetic: true },
});
policy = await client.activateGatePolicy(policy.id);
const decision = await client.recordDecision({
  agent_id: "synthetic-ts-agent",
  decision_type: "order_release",
  outcome: "approved",
  decided_at: occurredAt.toISOString(),
  policy_version: "quickstart-ts-1",
  metadata: { risk_level: "high", synthetic: true },
});
const providerRequest = {
  action: "synthetic.order.release",
  target_ref: "urn:lians:synthetic-order:SYNTHETIC-TS-001",
  decision_id: decision.id,
  arguments: { order_id: "SYNTHETIC-TS-001", synthetic: true },
};
const executionRequestHash = createHash("sha256")
  .update(canonicalJson(providerRequest))
  .digest("hex");
const verdict = await client.evaluateGate({
  action: "synthetic.order.release",
  target_ref: "urn:lians:synthetic-order:SYNTHETIC-TS-001",
  decision_id: decision.id,
  enforcement_principal_id: mediatorPrincipalId,
  permit_ttl_seconds: 30,
  execution_request_hash: executionRequestHash,
  risk_level: "high",
  policy_set_id: policy.id,
  context: { synthetic: true, recorder_run_id: recorderRunId },
});
const permit = verdict.execution_permit;
let permitConsumed = false;
if (permit && mediatorApiKey) {
  // A real mediator derives these claims from its normalized actual request
  // and recomputes the digest instead of trusting the evaluator's copy.
  const mediatorRequestHash = createHash("sha256")
    .update(canonicalJson(providerRequest))
    .digest("hex");
  const mediator = new LiansClient({ baseUrl, apiKey: mediatorApiKey });
  await mediator.consumeGateExecutionPermit({
    permit_id: permit.permit_id,
    token: permit.token,
    action: providerRequest.action,
    target_ref: providerRequest.target_ref,
    decision_id: providerRequest.decision_id,
    execution_request_hash: mediatorRequestHash,
  });
  permitConsumed = true;
}
console.log("gate", {
  principal: principal.principal_id,
  disposition: verdict.disposition,
  permit_id: permit?.permit_id,
  permit_consumed: permitConsumed,
});

const investigation = await client.createInvestigationCase({
  title: "Synthetic TypeScript control-plane exercise",
  severity: "low",
  gate_decision_id: verdict.id,
  metadata: { synthetic: true },
});
const task = await client.createRemediationTask(investigation.id, {
  expected_case_updated_at: investigation.updated_at,
  title: "Attest synthetic evidence review",
  owner_principal: principal.principal_id ?? undefined,
  metadata: { synthetic: true },
});
const taskInProgress = await client.updateRemediationTask(task.id, {
  expected_updated_at: task.updated_at,
  status: "in_progress",
});
const taskClosed = await client.closeRemediationTask(task.id, {
  expected_updated_at: taskInProgress.updated_at,
  statement: "Synthetic evidence review completed.",
  evidence_refs: [`gate:${verdict.id}`, `recorder-run:${recorderRunId}`],
});
const caseBeforeClose = await client.investigationCase(investigation.id);
const caseClosed = await client.closeInvestigationCase(investigation.id, {
  expected_updated_at: caseBeforeClose.updated_at,
  statement: "Synthetic exercise closed after the owned task was attested.",
  evidence_refs: [`attestation:${taskClosed.attestation.id}`],
  resolution_summary: "Synthetic control-plane path completed.",
});
console.log("investigation", { case_id: investigation.id, status: caseClosed.status });
}

main().catch((error: unknown) => {
  // The SDK never logs request bodies or credentials. Keep example failures terse.
  console.error(error instanceof Error ? error.message : "Synthetic quickstart failed");
  process.exitCode = 1;
});
